Django 2.1 compatibility

ransford opened the pull request:
Django 2.1 moved django.template.base.TOKEN_* into [an Enum class](https://github.com/django/django/blob/stable/2.1.x/django/template/base.py#L100) called django.template.base.TokenType.  Import TOKEN_VAR and TOKEN_BLOCK from that class instead of the old location if Django has version >= 2.1.

This may not be the ideal solution, but it got me up and running again.  I won't be offended if you close this PR in favor of a better approach!

ransford committed:
Django 2.1 compatibility

Django 2.1 moved django.template.base.TOKEN_* into an Enum class called
django.template.base.TokenType.  Import TOKEN_VAR and TOKEN_BLOCK from that
class instead of the old location if Django has version >= 2.1.

https://github.com/django/django/blob/stable/2.1.x/django/template/base.py#L100

ransford commented:
Any chance this will be merged? I believe it's a minor change.

vkurup commented:
Hi @ransford, sorry about the delay in response. We're not actively using this project in any client work at the moment. I'll try to get back up to speed on this shortly, but in the meantime, is this PR working for you? The change that you made looks reasonable to me, but the [Django PR](https://github.com/django/django/pull/9938/files) where this change was made also seems to change `token.token_type` from an integer to an Enum, which I think would mean we'd need to change line 124-133 as well.

That's based on just a quick read, so I could be way off, but I'll investigate...

ransford commented:
This patch is "working" (no sign of trouble) for me with Django 2.1 on my fork, but thanks for the more careful reading. I'll revisit this shortly in light of your suggestion.

ransford committed:
use enum type directly, rather than its .value

ransford committed:
add python3 x django 2.1 tox configuration

(Django 2.1 supports Python 3.{5,6,7})

ransford commented:
OK @vkurup, I *think* I've correctly addressed your comment, and I updated the tox configuration in 10042e9. I can't tell whether your CI mechanism picked up this PR, though!

vkurup commented:
Thanks @ransford! (I'll have to figure out the CI issues...)

I looked at all the previous code we had written using JSX and we never included anything inside the JSX tag that would hit either the `TOKEN_VAR` or `TOKEN_BLOCK` conditionals. I'm guessing that you don't either, since you didn't hit any problems with your PR. And the more I think about it, it doesn't make sense to allow, for example, a Django variable (like `{{ user_id }}`) inside of a JSX block. Because where would it get evaluated? If it got evaluated during the Django render, then the sha1 hash wouldn't match the value that `compilejsx` calculates. So, I think we should probably remove that bit of code and instead raise a warning about misuse if anything but `TOKEN_TEXT` is sent.

I'll merge this for now and create an issue to address future improvements.

Thanks (and thanks for your patience!)

ransford commented:
Fair enough! I didn't think I had tested very thoroughly. My app doesn't put anything other than JSX code in `{% jsx %}` blocks.

Thanks for this project!
